File: scripts/flutter/template/flutter_screen_template.py
#!/usr/bin/env python3
"""Flutter Screen 템플릿 생성 스크립트"""

import sys


def to_snake_case(name: str) -> str:
    """PascalCase를 snake_case로 변환"""
    result = []
    for i, char in enumerate(name):
        if char.isupper() and i > 0:
            result.append("_")
        result.append(char.lower())
    return "".join(result)


def generate_screen(screen_name: str) -> str:
    """Screen 템플릿 코드 생성"""

    # Screen 접미사가 없으면 추가
    if not screen_name.endswith("Screen"):
        screen_name = f"{screen_name}Screen"

    # Bloc 이름 생성 (Screen 접미사 제거)
    bloc_name = screen_name[:-6]  # 'Screen' 제거

    return f"""import 'package:flutter/material.dart';
import 'package:flutter_leaf_kit/flutter_leaf_kit.dart';

import '../bloc/{to_snake_case(bloc_name)}_bloc.dart';
import '../widgets/{to_snake_case(bloc_name)}_app_bar.dart';
import '../widgets/{to_snake_case(bloc_name)}_body_view.dart';

class {screen_name} extends LeafScreenStatefulWidget {{
  const {screen_name}({{super.key}});

  @override
  State<{screen_name}> createState() => _{screen_name}State();
}}

class _{screen_name}State extends LeafScreenState<{screen_name}> {{
  @override
  Color? get backgroundColor => null;

  @override
  Widget? buildScreen(BuildContext context) {{
    return BlocScreenConsumer<{bloc_name}Bloc, {bloc_name}State>(
      builder: (context, state) {{
        return buildScaffold(context, state);
      }},
      successListener: (context, state) {{}},
    );
  }}

  @override
  PreferredSizeWidget? buildAppBar(BuildContext context, Object? state) {{
    return const {bloc_name}AppBar();
  }}

  @override
  Widget buildBody(BuildContext context, Object? state) {{
    return const {bloc_name}BodyView();
  }}
}}
"""


def main():
    if len(sys.argv) < 2:
        print("Usage: screen_template.py <ScreenName>", file=sys.stderr)
        print("Example: screen_template.py Login", file=sys.stderr)
        sys.exit(1)

    screen_name = sys.argv[1]
    print(generate_screen(screen_name))


if __name__ == "__main__":
    main()

File: scripts/flutter/template/flutter-screen-template.mjs
#!/usr/bin/env node
// Flutter Screen 템플릿 생성 스크립트

import process from 'node:process';

function toSnakeCase(name) {
  let out = '';
  for (let i = 0; i < name.length; i++) {
    const ch = name[i];
    if (ch >= 'A' && ch <= 'Z' && i > 0) {
      out += '_';
    }
    out += ch.toLowerCase();
  }
  return out;
}

function generateScreen(screenNameIn, options = {}) {
  let screenName = screenNameIn;
  if (!screenName.endsWith('Screen')) {
    screenName = `${screenName}Screen`;
  }

  const blocName = screenName.slice(0, -6);
  const blocSnake = toSnakeCase(blocName);

  const { pkg, featureDir } = options;
  const useAbs = Boolean(pkg && featureDir);
  const blocImport = useAbs
    ? `package:${pkg}/features/${featureDir}/presentation/bloc/${blocSnake}_bloc.dart`
    : `../bloc/${blocSnake}_bloc.dart`;
  const appBarImport = useAbs
    ? `package:${pkg}/features/${featureDir}/presentation/views/${blocSnake}_app_bar.dart`
    : `../views/${blocSnake}_app_bar.dart`;
  const bodyImport = useAbs
    ? `package:${pkg}/features/${featureDir}/presentation/views/${blocSnake}_body_view.dart`
    : `../views/${blocSnake}_body_view.dart`;

  return `import 'package:flutter/material.dart';
import 'package:flutter_leaf_kit/flutter_leaf_kit.dart';

import '${blocImport}';
import '${appBarImport}';
import '${bodyImport}';

class ${screenName} extends LeafScreenStatefulWidget {
  const ${screenName}({super.key});

  @override
  State<${screenName}> createState() => _${screenName}State();
}

class _${screenName}State extends LeafScreenState<${screenName}> {
  @override
  Color? get backgroundColor => null;

  @override
  Widget? buildScreen(BuildContext context) {
    return BlocScreenConsumer<${blocName}Bloc, ${blocName}State>(
      builder: buildScaffold,
      successListener: (context, state) {},
    );
  }

  @override
  PreferredSizeWidget? buildAppBar(BuildContext context, Object? state) {
    return const ${blocName}AppBar();
  }

  @override
  Widget buildBody(BuildContext context, Object? state) {
    return const ${blocName}BodyView();
  }
}
`;
}

function parseArgs(argv) {
  const opts = { screenName: '', pkg: '', featureDir: '' };
  const rest = [...argv];
  while (rest.length > 0) {
    const a = rest.shift();
    if (a === '--package' || a === '-p') {
      opts.pkg = rest.shift() ?? '';
    } else if (a === '--feature-dir' || a === '-f') {
      opts.featureDir = rest.shift() ?? '';
    } else if (a === '-h' || a === '--help') {
      printUsage();
      process.exit(0);
    } else if (a.startsWith('-')) {
      process.stderr.write(`Unknown option: ${a}\n`);
      process.exit(2);
    } else if (!opts.screenName) {
      opts.screenName = a;
    } else {
      process.stderr.write(`Unexpected argument: ${a}\n`);
      process.exit(2);
    }
  }
  return opts;
}

function printUsage() {
  process.stderr.write('Usage: flutter-screen-template.mjs <ScreenName> [--package <pkg>] [--feature-dir <dir>]\n');
  process.stderr.write('Example: flutter-screen-template.mjs Login --package vocabit_app --feature-dir login\n');
}

function main() {
  const opts = parseArgs(process.argv.slice(2));
  if (!opts.screenName) {
    printUsage();
    process.exit(1);
  }
  console.log(generateScreen(opts.screenName, { pkg: opts.pkg, featureDir: opts.featureDir }));
}

main();
